File: packages/typia/src/transformers/features/json/JsonSchemaTransformer.ts
import ts from "typescript";

import { LiteralFactory } from "../../../factories/LiteralFactory";
import { MetadataCollection } from "../../../factories/MetadataCollection";
import { MetadataFactory } from "../../../factories/MetadataFactory";

import { Metadata } from "../../../schemas/metadata/Metadata";

import { JsonSchemaProgrammer } from "../../../programmers/json/JsonSchemaProgrammer";
import { JsonSchemasProgrammer } from "../../../programmers/json/JsonSchemasProgrammer";

import { ValidationPipe } from "../../../typings/ValidationPipe";

import { IJsonSchemaUnit } from "../../../module";
import { ITransformProps } from "../../ITransformProps";
import { TransformerError } from "../../TransformerError";

export namespace JsonSchemaTransformer {
  export const transform = (
    props: Pick<ITransformProps, "context" | "expression">,
  ): ts.Expression => {
    if (!props.expression.typeArguments?.length)
      throw new TransformerError({
        code: "typia.json.schema",
        message: "no generic argument.",
      });

    //----
    // GET ARGUMENTS
    //----
    // VALIDATE TUPLE ARGUMENTS
    const top: ts.Node | undefined = props.expression.typeArguments[0];
    if (top === undefined || ts.isTypeNode(top) === false)
      return props.expression;

    // GET TYPES
    const type: ts.Type = props.context.checker.getTypeFromTypeNode(top);
    if (type.isTypeParameter())
      throw new TransformerError({
        code: "typia.json.schema",
        message: "non-specified generic argument.",
      });

    // ADDITIONAL PARAMETERS
    const version: "3.0" | "3.1" = getParameter<"3.0" | "3.1">({
      checker: props.context.checker,
      name: "Version",
      is: (str) => str === "3.0" || str === "3.1",
      cast: (str) => str as "3.0" | "3.1",
      default: () => "3.1",
    })(props.expression.typeArguments[1]);

    //----
    // GENERATORS
    //----
    // METADATA
    const analyze = (validate: boolean): Metadata => {
      const results: ValidationPipe<Metadata, MetadataFactory.IError> =
        MetadataFactory.analyze({
          checker: props.context.checker,
          transformer: props.context.transformer,
          options: {
            absorb: validate,
            constant: true,
            escape: true,
            validate:
              validate === true ? JsonSchemasProgrammer.validate : undefined,
          },
          collection: new MetadataCollection({
            replace: MetadataCollection.replace,
          }),
          type,
        });
      if (results.success === false)
        throw TransformerError.from({
          code: "typia.json.schema",
          errors: results.errors,
        });
      return results.data;
    };
    analyze(true);

    // APPLICATION
    const app: IJsonSchemaUnit<any> = JsonSchemaProgrammer.write({
      version,
      metadata: analyze(false),
    });
    return ts.factory.createAsExpression(
      LiteralFactory.write(app),
      props.context.importer.type({
        file: "typia",
        name: "IJsonSchemaUnit",
        arguments: [
          ts.factory.createLiteralTypeNode(
            ts.factory.createStringLiteral(version),
          ),
        ],
      }),
    );
  };

  const getParameter =
    <Value>(props: {
      checker: ts.TypeChecker;
      name: string;
      is: (value: string) => boolean;
      cast: (value: string) => Value;
      default: () => Value;
    }) =>
    (node: ts.TypeNode | undefined): Value => {
      if (!node) return props.default();

      // CHECK LITERAL TYPE
      const type: ts.Type = props.checker.getTypeFromTypeNode(node);
      if (
        !type.isLiteral() &&
        (type.getFlags() & ts.TypeFlags.BooleanLiteral) === 0
      )
        throw new TransformerError({
          code: "typia.json.schema",
          message: `generic argument "${props.name}" must be constant.`,
        });

      // GET VALUE AND VALIDATE IT
      const value = type.isLiteral()
        ? type.value
        : props.checker.typeToString(type);
      if (typeof value !== "string" || props.is(value) === false)
        throw new TransformerError({
          code: "typia.json.schema",
          message: `invalid value on generic argument "${props.name}".`,
        });
      return props.cast(value);
    };
}

File: packages/typia/native/transform/features/json/JsonSchemaTransformer.go
package json

import (
  "fmt"

  shimast "github.com/microsoft/typescript-go/shim/ast"
  shimchecker "github.com/microsoft/typescript-go/shim/checker"
  nativefactories "github.com/samchon/typia/packages/typia/native/core/factories"
  nativejsonprogrammers "github.com/samchon/typia/packages/typia/native/core/programmers/json"
  schemametadata "github.com/samchon/typia/packages/typia/native/core/schemas/metadata"
  nativetransform "github.com/samchon/typia/packages/typia/native/transform/internal"
)

type jsonSchemaTransformerNamespace struct{}

var JsonSchemaTransformer = jsonSchemaTransformerNamespace{}

func (jsonSchemaTransformerNamespace) Transform(props nativetransform.ITransformProps) *shimast.Node {
  if props.Expression == nil || props.Expression.TypeArguments == nil || len(props.Expression.TypeArguments.Nodes) == 0 {
    panic(nativetransform.NewTransformerError(nativetransform.TransformerError_IProps{
      Code:    "typia.json.schema",
      Message: "no generic argument.",
    }))
  }

  top := props.Expression.TypeArguments.Nodes[0]
  if top == nil {
    return props.Expression.AsNode()
  }

  typ := props.Context.Checker.GetTypeFromTypeNode(top)
  if typ != nil && typ.IsTypeParameter() {
    panic(nativetransform.NewTransformerError(nativetransform.TransformerError_IProps{
      Code:    "typia.json.schema",
      Message: "non-specified generic argument.",
    }))
  }

  var versionNode *shimast.Node
  if len(props.Expression.TypeArguments.Nodes) > 1 {
    versionNode = props.Expression.TypeArguments.Nodes[1]
  }
  version := jsonTransformer_getParameter(jsonTransformer_getParameterProps{
    Checker: props.Context.Checker,
    Code:    "typia.json.schema",
    Name:    "Version",
    Is: func(str string) bool {
      return str == "3.0" || str == "3.1"
    },
    Default: func() string {
      return "3.1"
    },
  })(versionNode)

  analyze := func(validate bool) *schemametadata.MetadataSchema {
    var validator nativefactories.MetadataFactory_Validator
    if validate {
      validator = jsonTransformer_schemasValidator
    }
    result := nativefactories.MetadataFactory.Analyze(nativefactories.MetadataFactory_IProps{
      Checker:     props.Context.Checker,
      Transformer: props.Context.Transformer,
      Options: nativefactories.MetadataFactory_IOptions{
        Absorb:   validate,
        Constant: true,
        Escape:   true,
        Validate: validator,
      },
      Components: schemametadata.NewMetadataCollection(&schemametadata.MetadataCollection_IOptions{
        Replace: schemametadata.MetadataCollection_replace,
      }),
      Type: typ,
    })
    if result.Success == false {
      panic(nativetransform.TransformerError_from(struct {
        Code   string
        Errors []nativetransform.TransformerError_MetadataFactory_IError
      }{
        Code:   "typia.json.schema",
        Errors: jsonTransformer_errors(result.Errors),
      }))
    }
    return result.Data
  }
  analyze(true)

  return nativejsonprogrammers.JsonSchemaProgrammer.Write(nativejsonprogrammers.JsonSchemaProgrammer_IWriteProps{
    Context:  props.Context,
    Version:  version,
    Metadata: analyze(false),
  })
}

type jsonTransformer_getParameterProps struct {
  Checker *shimchecker.Checker
  Code    string
  Name    string
  Is      func(value string) bool
  Default func() string
}

func jsonTransformer_getParameter(props jsonTransformer_getParameterProps) func(node *shimast.Node) string {
  return func(node *shimast.Node) string {
    if node == nil {
      return props.Default()
    }
    typ := props.Checker.GetTypeFromTypeNode(node)
    if typ == nil || (typ.Flags()&shimchecker.TypeFlagsLiteral) == 0 {
      panic(nativetransform.NewTransformerError(nativetransform.TransformerError_IProps{
        Code:    props.Code,
        Message: fmt.Sprintf("generic argument \"%s\" must be constant.", props.Name),
      }))
    }

    value := ""
    if typ.Flags()&shimchecker.TypeFlagsBooleanLiteral != 0 {
      value = props.Checker.TypeToString(typ)
    } else {
      if str, ok := typ.AsLiteralType().Value().(string); ok {
        value = str
      }
    }
    if value == "" || props.Is(value) == false {
      panic(nativetransform.NewTransformerError(nativetransform.TransformerError_IProps{
        Code:    props.Code,
        Message: fmt.Sprintf("invalid value on generic argument \"%s\".", props.Name),
      }))
    }
    return value
  }
}

func jsonTransformer_schemasValidator(props struct {
  Metadata *schemametadata.MetadataSchema
  Explore  nativefactories.MetadataFactory_IExplore
  Top      *schemametadata.MetadataSchema
}) []string {
  return nativejsonprogrammers.JsonSchemasProgrammer.Validate(struct {
    Metadata *schemametadata.MetadataSchema
    Explore  nativefactories.MetadataFactory_IExplore
  }{
    Metadata: props.Metadata,
    Explore:  props.Explore,
  })
}

func jsonTransformer_applicationValidator(props struct {
  Metadata *schemametadata.MetadataSchema
  Explore  nativefactories.MetadataFactory_IExplore
  Top      *schemametadata.MetadataSchema
}) []string {
  return nativejsonprogrammers.JsonApplicationProgrammer.Validate(struct {
    Metadata *schemametadata.MetadataSchema
    Explore  nativefactories.MetadataFactory_IExplore
  }{
    Metadata: props.Metadata,
    Explore:  props.Explore,
  })
}

func jsonTransformer_errors(errors []nativefactories.MetadataFactory_IError) []nativetransform.TransformerError_MetadataFactory_IError {
  output := make([]nativetransform.TransformerError_MetadataFactory_IError, 0, len(errors))
  for _, err := range errors {
    output = append(output, nativetransform.TransformerError_MetadataFactory_IError{
      Name: err.Name,
      Explore: nativetransform.TransformerError_MetadataFactory_IExplore{
        Object:    err.Explore.Object,
        Property:  err.Explore.Property,
        Parameter: err.Explore.Parameter,
        Output:    err.Explore.Output,
      },
      Messages: append([]string{}, err.Messages...),
    })
  }
  return output
}
